components/Article.js

```diff
@@ -16,6 +18,12 @@ export default function Article({
 }) {
   const isAmp = useAmp();
 
+  useEffect(() => {
+    // this is used for the canonical link tag in the Layout component
+    let canonicalArticleUrl = generateArticleUrl(window.location.href, article);
+    siteMetadata['canonicalUrl'] = canonicalArticleUrl;
+  }, []);
+
```

Comment:
I think we shouldn't use `useEffect` or `window` here. We don't want this code to run only on client-side renders, because that means the initial baked version of the site (what crawlers will see) will have the incorrect value, because this code doesn't run on the server-side render.

I'd recommend getting the base URL some other way so we can use it on the server-side render. Maybe that should be a `siteMetadata` property or env var?

Reply:
good point - I forgot that we also had both an env var and a metadata field for this base url of the site. I decided to go with both: you can override the url with an env var (useful for local debugging, at least), fallback to the metadata value.

this requires you update your local .env to include one extra entry, a NEXT_PUBLIC_SITE_URL. Mine looks like this:

```
SITE_URL=http://localhost:3000
NEXT_PUBLIC_SITE_URL=http://localhost:3000
```